### PR TITLE
change bfs to bibfs in allpath

### PR DESCRIPTION
**Pull request by nevermore3**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [x] enhancement

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

**Review by yixinglu — DISMISSED** [review]

**Review by yixinglu — DISMISSED** [review]

File: src/graph/executor/algo/AllPathsExecutor.cpp [hunk 1]
```diff
@@ -158,72 +158,21 @@ folly::Future<Status> AllPathsExecutor::getNeighbors(bool reverse) {
...
         auto key = folly::sformat("buildAdjTime {}step[{}]", reverse ? "reverse " : "", step);
-        if (reverse) {
-          rightNextStepVids_.clear();
-          expandFromRight(iter.get());
-        } else {
-          leftNextStepVids_.clear();
-          expandFromLeft(iter.get());
-        }
+        expand(iter.get(), reverse);
+        statsLock_.lock();
```
Comment (codesigner) [inline src/graph/executor/algo/AllPathsExecutor.cpp:162]: use RAII, what if emplace exceed memory?

File: src/graph/executor/algo/AllPathsExecutor.cpp [hunk 2]
```diff
@@ -428,28 +320,27 @@ AllPathsExecutor::doBuildPath(size_t step,
...
+      return doBuildPath(step + 1, 0, newPathsSize, newPathsPtr, reverse);
     }));
   } else {
     for (size_t _start = 0; _start < newPathsSize; _start += FLAGS_path_batch_size) {
       auto tmp = _start + FLAGS_path_batch_size;
       auto _end = tmp > newPathsSize ? newPathsSize : tmp;
-      futures.emplace_back(folly::via(runner(), [this, step, _start, _end, newPathsPtr]() {
-        return doBuildPath(step + 1, _start, _end, newPathsPtr);
+      futures.emplace_back(folly::via(runner(), [this, step, _start, _end, newPathsPtr, reverse]() {
+        return doBuildPath(step + 1, _start, _end, newPathsPtr, reverse);
```
Comment (codesigner) [inline src/graph/executor/algo/AllPathsExecutor.cpp:336]: not familiar with the logic here, should  _start  also be updated?

File: src/graph/executor/algo/AllPathsExecutor.cpp [hunk 3]
```diff
@@ -462,6 +353,241 @@ AllPathsExecutor::doBuildPath(size_t step,
...
+  for (auto& vid : initVids) {
+    auto findVid = hashTable_.find(vid);
+    if (findVid == hashTable_.end()) {
+      continue;
+    }
+    auto& paths = findVid->second;
+    Value emptyPropVertex(Vertex(vid, {}));
+    if (!reverse) {
+      for (auto& path : paths) {
+        ++cnt_;
```
Comment (codesigner) [inline src/graph/executor/algo/AllPathsExecutor.cpp:428]: defalt ++ is memory_order_seq_cst, is it what you want, or here we can use memory_order_relaxed?

File: src/graph/executor/algo/AllPathsExecutor.h [hunk 4]
```diff
@@ -107,8 +125,12 @@ class AllPathsExecutor final : public StorageAccessExecutor {
 
   DataSet result_;
   std::vector<Value> emptyPropVids_;
+  std::unordered_multiset<Value, VertexHash, VertexEqual> emptyPropVertices_;
```
Comment (codesigner) [inline src/graph/executor/algo/AllPathsExecutor.h:128]: Is it enough to just store VertexId for this emptyPropVertices set?

File: src/graph/executor/algo/AllPathsExecutor.cpp (hunk 2 above)
Comment (nevermore3) [reply src/graph/executor/algo/AllPathsExecutor.cpp:336]: each iteration _start in the for statement will be updated

**Review by codesigner — COMMENTED** [review]
Have you tested imporvement on a decent dataset,  compare bibfs to bfs.

**Comment by nevermore3** [comment]
> Have you tested imporvement on a decent dataset, compare bibfs to bfs.

 BFS
image

Bidirectional BFS
image


this is BFS search space
image

this is BiBFS search space
image

**Review by codesigner — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Pull request merged**